File: PCPKotlinExercises/src/main/java/ch/hslu/pcp/kotlin/ex/original/Exercise3.java
package ch.hslu.pcp.kotlin.ex.original;

import java.util.concurrent.CompletableFuture;

public final class Exercise3 {

    public static void main(String[] args) {
        final CompletableFuture<Long> longLastingTaskFuture = CompletableFuture.supplyAsync(() -> WaitPrintAndReturnX(3000L));
        final CompletableFuture<Long> evenLongerLastingTaskFuture = CompletableFuture.supplyAsync(() -> WaitPrintAndReturnX(6000L));
        final CompletableFuture<String> lastTaskFuture = longLastingTaskFuture.thenCombineAsync(evenLongerLastingTaskFuture, (s, t) -> {
            try {
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            return "was waiting for " + (s + t + 2000) + "ms";
        });
        lastTaskFuture.thenAccept(System.out::print);
        System.out.println("-> Now waiting for things to happen!");
        for (int i = 0; i < 20; i++) {
            System.out.print(".");
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        System.out.println("\n-> Done.");
    }

    private static long WaitPrintAndReturnX(final long value) {
        try {
            Thread.sleep(value);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.print(value);
        return value;
    }
}

File: PCPKotlinExercises/src/main/java/ch/hslu/pcp/kotlin/ex/solution/Exercise3.kt
package ch.hslu.pcp.kotlin.ex.solution

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlin.coroutines.EmptyCoroutineContext

private fun waitPrintAndReturnX(value: Long): Long {
    try {
        Thread.sleep(value)
    } catch (e: InterruptedException) {
        e.printStackTrace()
    }
    print(value)
    return value
}

fun main(args: Array<String>) {
    val longLastingTaskFuture= CoroutineScope(EmptyCoroutineContext).async { waitPrintAndReturnX(3000) }
    val evenLongerLastingTaskFuture= CoroutineScope(EmptyCoroutineContext).async { waitPrintAndReturnX(6000) }
    CoroutineScope(EmptyCoroutineContext).async {
        val val1 = longLastingTaskFuture.await()
        val val2 = evenLongerLastingTaskFuture.await()
        try {
            Thread.sleep(2000)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        print("was waiting for " + (val1 + val2 + 2000) + "ms")
    }
    println("-> Now waiting for things to happen!")
    for (i in 0..19) {
        print(".")
        try {
            Thread.sleep(500)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }
    println("\n-> Done.")
}